components/operations-controller/controllers/operation_controller.go

```diff
@@ -16,36 +16,428 @@ package controllers
+	if !isWebhookTimeoutReached(r.webhookConfig, webhook, operation.CreationTimestamp.Time) {
+		requeueAfter := r.webhookConfig.RequeueInterval
+		if webhook.RetryInterval != nil {
+			requeueAfter = time.Duration(*webhook.RetryInterval)
+		}
+		return ctrl.Result{RequeueAfter: requeueAfter}, nil
+	}
+	return r.updateStatusWithError(ctx, logger, operation, webhookErr)
+}
+
+func parseRequestData(operation *v1alpha1.Operation) (webhook.RequestData, error) {
+	str := struct {
+		Application graphql.Application
+		TenantID    string
+		Headers     map[string]string
+	}{}
+
+	if err := json.Unmarshal([]byte(operation.Spec.RequestData), &str); err != nil {
+		return webhook.RequestData{}, err
+	}
+
+	return webhook.RequestData{
+		Application: &str.Application,
+		TenantID:    str.TenantID,
+		Headers:     str.Headers,
+	}, nil
+}
+
+func prepareDirectorRequest(operation v1alpha1.Operation) director.Request {
+	return prepareDirectorRequestWithError(operation, nil)
+}
+
+func prepareDirectorRequestWithError(operation v1alpha1.Operation, err error) director.Request {
+	request := director.Request{
+		OperationType: graphql.OperationType(operation.Spec.OperationType),
+		ResourceType:  resource.Type(operation.Spec.ResourceType),
+		ResourceID:    operation.Spec.ResourceID,
+	}
+
+	if err != nil {
+		request.Error = err.Error()
+	}
+
+	return request
+}
+
+func isReconciliationTimeoutReached(operation *v1alpha1.Operation, reconciliationTimeout time.Duration) bool {
+	operationEndTime := operation.ObjectMeta.CreationTimestamp.Time.Add(reconciliationTimeout)
+
+	return time.Now().After(operationEndTime)
+}
+
+func isWebhookTimeoutReached(cfg *web_hook.Config, webhook graphql.Webhook, creationTime time.Time) bool {
+	webhookTimeout := cfg.WebhookTimeout
+	if webhook.Timeout != nil {
+		webhookTimeout = time.Duration(*webhook.Timeout)
+	}
+
+	operationEndTime := creationTime.Add(webhookTimeout)
+	return time.Now().After(operationEndTime)
+}
+
+func getNextPollTime(webhook *graphql.Webhook, webhookStatus v1alpha1.Webhook, timeLayout string) (time.Duration, error) {
+	if webhookStatus.LastPollTimestamp == "" {
+		return 0, nil
+	}
+
+	lastPollTimestamp, err := time.Parse(timeLayout, webhookStatus.LastPollTimestamp)
+	if err != nil {
+		return 0, err
+	}
+
+	nextPollTime := lastPollTimestamp.Add(time.Duration(*webhook.RetryInterval) * time.Second)
+	return nextPollTime.Sub(time.Now()), nil
+}
+
+func setDefaultStatus(operation v1alpha1.Operation) *v1alpha1.Operation {
+	overrideStatus := false
+	if operation.Generation != operation.Status.ObservedGeneration {
+		overrideStatus = true
+		operation.Status.ObservedGeneration = operation.Generation
+		operation.Status.Phase = ""
+	}
+
+	if operation.Status.Conditions == nil || overrideStatus {
+		operation.Status.Conditions = make([]v1alpha1.Condition, 0)
+	}
+
+	hasReady := false
+	hasError := false
+	for _, condition := range operation.Status.Conditions {
+		if condition.Type == v1alpha1.ConditionTypeReady {
```

Comment:
its a common pattern to define some utils for working with conditions, otherwise its too verbose and hard to follow.. Getting a condition by type, Initializing conditions to default values, Setting a condition by type and so on.. 

In general a well written loop likes like a "sentence" - basically most if checks are outsourced to smaller methods so it can be a readable thing in the end, without caring too much about the details. Example https://github.com/kyma-project/kyma/blob/132d719c59388fc01af7e8b382f3e65dbf346dee/components/kyma-operator/pkg/k8s/controller.go. There are probably other good examples, havent looked too much yet, I can do if you need further details

Reply:
Thanks for the feedback - I introduced a `StatusManager` abstraction to hide all the details regarding maintaining an Operation CR's status. Also the `setDefaultStatus` method was removed.

Reply:
Also in general I tried streamlining the main loop logic to leverage the different abstractions that we have so that the whole loop reads easier. The main loop function has reduced more than twice in terms of code lines as a result, but of course there are limits to how much we can shorten it as the reconcile loop has some corner case branches which need addressing and which cannot be covered cleanly by abstractions (ex. invalid requestObject, deriving the correct webhook entity from the retrieved application, calculating next poll time.. to name a few).